include/sparrow/mp_utils.hpp

```diff
@@ -14,19 +14,228 @@
 
 #pragma once
 
+#include <type_traits>
 #include <iterator>
 
-namespace sparrow
+namespace sparrow::mpl
 {
-    namespace mpl
+    /// A sequence of types, used for meta-programming operations.
+    template< class... T >
+    struct typelist {};
+
+    /// @returns The count of types contained in a given `typelist`.
+    template< class... T >
+    constexpr std::size_t size(typelist<T...>) { return sizeof...(T); }
+
+    template< class L, template<class...> class U >
+    struct is_type_instance_of : std::false_type {};
+
+    template< template<class...> class L, template<class...> class U, class... T>
+        requires std::is_same_v< L<T...>, U<T...> >
+    struct is_type_instance_of< L<T...>, U > : std::true_type {};
+
+    /// `true` if `T` is a concrete type template instanciation of `U` which is a type template.
+    /// Example: is_type_instant_of_v< std::vector<int>, std::vector > == true
+    template< class T, template< class... > class U >
+    constexpr bool is_type_instance_of_v = is_type_instance_of<T, U>::value;
+
+    /// Matches any type which is an instance of `typelist`.
+    /// Examples:
+    ///     static_assert(any_typelist< typelist<int, float, std::vector<int> > > == true);
+    ///     static_assert(any_typelist< std::vector<int> > == false);
+    template< typename TList >
+    concept any_typelist = is_type_instance_of_v<TList, typelist>;
+
+    // TODO: document this!
+    template <class T, bool is_const>
+    struct constify
+        : std::conditional<is_const, const T, T>
+    {
+    };
+
+    template <class T, bool is_const>
+    using constify_t = typename constify<T, is_const>::type;
```

Comment:
```suggestion
    template <class T, bool is_const>
    struct constify
        : std::conditional<is_const, const T, T>
    {
    };

    // `constify_t` is required since `std::add_const_t<T&>`
    // is not `const T&` but `T&`.
    template <class T, bool is_const>
    using constify_t = typename constify<T, is_const>::type;
```